arch/x86/core/tls.c

```diff
@@ -0,0 +1,45 @@
+/*
+ * Copyright (c) 2020 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <kernel.h>
+#include <kernel_structs.h>
+#include <kernel_internal.h>
+#include <kernel_tls.h>
+#include <sys/util.h>
+
+size_t arch_tls_stack_setup(struct k_thread *new_thread, char *stack_ptr)
+{
+	/*
+	 * TLS area for x86 and x86_64 has the data/bss first,
+	 * then a pointer pointing to itself. The address to
+	 * this pointer needs to be stored in register GS (x86)
+	 * or FS (x86_64). GCC generates code which reads this
+	 * pointer and offsets from this pointer are used to
+	 * access data.
+	 */
+
+	uintptr_t *self_ptr;
+
+	/*
+	 * Since we are populating things backwards, store
+	 * the pointer to the TLS area at top of stack.
+	 */
+	stack_ptr -= sizeof(uintptr_t);
+	self_ptr = (void *)stack_ptr;
+	*self_ptr = POINTER_TO_UINT(stack_ptr);
+
+	/*
+	 * Set thread TLS pointer as this is used to populate
+	 * FS/GS at context switch.
+	 */
+	new_thread->tls = POINTER_TO_UINT(self_ptr);
+
+	/* Setup the TLS data */
+	stack_ptr -= z_tls_data_size();
```

Comment:
Do you want to align this to a 16-byte boundary as recommended by the architecture reference?

__Segment base addresses should be aligned to 16-byte boundaries. Although 16-byte alignment is not required, this alignment allows programs to maximize performance by aligning code and data on 16-byte boundaries.__

Reply:
Come to think of it... the segment base address (in this case `%gs`) is strictly used to locate the pointer (which points to itself). So there is no need to align on 16-byte, but probably need to be 4-byte aligned, which, given how `arch_tls_stack_setup()` is being called by `setup_thread_stack()`, should have been aligned correctly.